2022Swerve/src/main/java/frc/robot/subsystems/SwerveSubsystem.java

```diff
@@ -0,0 +1,82 @@
+package frc.robot.subsystems;
+
+import com.kauailabs.navx.frc.AHRS;
+
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.kinematics.SwerveDriveKinematics;
+import edu.wpi.first.math.kinematics.SwerveModuleState;
+import edu.wpi.first.wpilibj.SPI;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj2.command.InstantCommand;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import edu.wpi.first.wpilibj2.command.WaitUntilCommand;
+import frc.robot.Constants.DriveConstants;
+import frc.robot.hardware.AbsoluteEncoders;
+import frc.robot.hardware.MotorControllers;
+
+public class SwerveSubsystem extends SubsystemBase{
+    private final SwerveModule frontLeft = new SwerveModule(
+        MotorControllers.FrontLeftModuleDrive.getMotor(),
+        MotorControllers.FrontLeftModuleTurn.getMotor(),
+        AbsoluteEncoders.FrontLeftModule.getSupplier());
```

Comment:
Trying to think through this.

getSupplier gets a supplier that gets the turn in radians.
We need a supplier here because the SwerveModule is using a supplier.

You could have also chosen to do:
() -> AbsoluteEncoders.FrontLeftModule.calculateAngleRadians(),

We should talk more about what the benefit is in doing it this way.

Reply:
`() -> AbsoluteEncoders.FrontLeftModule.calculateAngleRadians(),`
What exactly is the lambda doing here?

This strikes me as unnecessarily verbose, especially to be called every time an angle calculation is needed. In the current implementation, it is only necessary the line `AbsoluteEncoders.FrontLeftModule.getSupplier(),` is called when instantiating the supplier variable. From that point onward, it is only necessary that encoderSupplier.get() is called, which is much less verbose.

Reply:
>   From that point onward, it is only necessary that encoderSupplier.get() is called, which is much less verbose.
I don't see encoderSupplier here, so assume that you mean something like this:
encoderSupplier = AbsoluteEncoders.FrontLeftModule.getSupplier();
// Call this anywhere in the code afterwards
encoderSupplier.get();

I am saying that you can make the following choice as well:
encoderSupplier = () -> AbsoluteEncoders.FrontLeftModule.calculateAngleRadians();
// Call this anywhere in the code afterwards
encoderSupplier.get();

Reply:
> I don't see encoderSupplier here

Example object name

>I am saying that you can make the following choice as well:
encoderSupplier = () -> AbsoluteEncoders.FrontLeftModule.calculateAngleRadians();
// Call this anywhere in the code afterwards
encoderSupplier.get();

Does .get() work on lambdas like this? Or would you just call encoderSupplier?

Reply:
This is my first time working with Supplier's.  From what I understand, if encoderSupplier is defined as:
Supplier<double> encoderSupplier;

Then yeah, the code above should work.  ( Javadoc doesn't have an example, but the following has an example that is exactly like that. )
https://www.geeksforgeeks.org/supplier-interface-in-java-with-examples/

Reply:
I do like the use of a lambda here, but I have a single aversion to using it; With my use of a supplier, the class is easier to understand and implement within one's own subsystem, but with the lambda, which provides minimal benefit, one can no longer understand the intended implementation by reading the class by itself, and could resort to using the class without the best practices (i.e. verbosely calling the calculate method every time an encoder value is needed).